diag_manager/fms_diag_yaml.F90

```diff
@@ -1,62 +1,324 @@
-!type(diag_fields_type)function get_diag_table_field (field_name) result (field)
-! character(len=*), intent(IN) :: field_name
-! integer :: i
-! do i = 1,size(diag_fields)
-!     if (trim(field_name) == trim(fms_c2f_string(diag_fields(i)%fname))) then
-!          field = diag_fields(i)
-!write (6,*) field_name//" Found"
-!
-!          return
-!     endif
-! enddo
-! field = null_field_type
-!
-!end function get_diag_table_field
+!> @brief Uses the yaml_parser_mod to read in the diag_table and fill in the
+!! diag_yaml object
+subroutine diag_yaml_object_init
+  integer              :: diag_yaml_id !< Id for the diag_table yaml
+  integer              :: nfiles       !< Number of files in the diag_table yaml
+  integer, allocatable :: file_ids(:)  !< Ids of the files in the diag_table yaml
+  integer              :: i, j         !< For do loops
+  integer              :: total_nvars  !< The total number of variables in the diag_table yaml
+  integer              :: var_count    !< The current number of variables added to the diag_yaml obj
+  integer              :: nvars        !< The number of variables in the current file
+  integer, allocatable :: var_ids(:)   !< Ids of the variables in diag_table yaml
+
+  diag_yaml_id = open_and_parse_file("diag_table.yaml")
+
+  call diag_get_value_from_key(diag_yaml_id, 0, "title", diag_yaml%diag_title)
+  call get_value_from_key(diag_yaml_id, 0, "base_date", diag_yaml%diag_basedate)
+
+  nfiles = get_num_blocks(diag_yaml_id, "diag_files")
+  allocate(diag_yaml%diag_files(nfiles))
+  allocate(file_ids(nfiles))
+  call get_block_ids(diag_yaml_id, "diag_files", file_ids)
+
+  total_nvars = get_total_num_vars(diag_yaml_id, file_ids)
+  allocate(diag_yaml%diag_fields(total_nvars))
+
+  var_count = 0
+  do i = 1, nfiles
+    call diag_yaml_files_obj_init(diag_yaml%diag_files(i))
+    call fill_in_diag_files(diag_yaml_id, file_ids(i), diag_yaml%diag_files(i))
+
+    nvars = 0
+    nvars = get_num_blocks(diag_yaml_id, "varlist", parent_block_id=file_ids(i))
+    allocate(var_ids(nvars))
+    call get_block_ids(diag_yaml_id, "varlist", var_ids, parent_block_id=file_ids(i))
+    do j = 1, nvars
+      var_count = var_count + 1
+      call fill_in_diag_fields(diag_yaml_id, var_ids(j), diag_yaml%diag_fields(var_count))
+    enddo
+    deallocate(var_ids)
+  enddo
+
+  deallocate(file_ids)
+end subroutine
+
+!> @brief Destroys the diag_yaml object
+subroutine diag_yaml_object_end()
+  integer :: i !< For do loops
+
+  do i = 1, size(diag_yaml%diag_files, 1)
+    if(allocated(diag_yaml%diag_files(i)%file_global_meta)) deallocate(diag_yaml%diag_files(i)%file_global_meta)
+  enddo
+  if(allocated(diag_yaml%diag_files)) deallocate(diag_yaml%diag_files)
+
+  do i = 1, size(diag_yaml%diag_fields, 1)
+    if(allocated(diag_yaml%diag_fields(i)%var_attributes)) deallocate(diag_yaml%diag_fields(i)%var_attributes)
+  enddo
+  if(allocated(diag_yaml%diag_fields)) deallocate(diag_yaml%diag_fields)
```

Comment:
This looks like work that could be shared between 2 threads because one loop doesn't depend on the other.